### PR TITLE
Fix Expr(:loopinfo) codegen

### LINKED ISSUE #50660: `julia.loopinfo_marker` can be attaced to wrong loop.
Noticed by @lcw in https://github.com/JuliaGPU/KernelAbstractions.jl/issues/411

Using the snippet in the PR above we see:

```
➜  julia JULIA_LLVM_ARGS="-debug-only=lower_simd_loop" ./julia --project=~/src/KernelAbstractions ~/src/KernelAbstractions/codegen.jl
LSL: loopinfo marker found
LSL: loop header: 
L2:                                               ; preds = %L2, %top
  %value_phi = phi i64 [ 1, %top ], [ %7, %L2 ]
  %2 = sitofp i64 %value_phi to double, !dbg !30
  %3 = add i64 %value_phi, -1, !dbg !37
  %4 = bitcast {} addrspace(10)* %"a::Array" to double addrspace(13)* addrspace(10)*, !dbg !37
  %5 = addrspacecast double addrspace(13)* addrspace(10)* %4 to double addrspace(13)* addrspace(11)*, !dbg !37
  %.data10 = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %5, align 8, !dbg !37, !tbaa !46, !alias.scope !49, !noalias !52, !nonnull !11
  %6 = getelementptr inbounds double, double addrspace(13)* %.data10, i64 %3, !dbg !37
  store double %2, double addrspace(13)* %6, align 8, !dbg !37, !tbaa !57, !alias.scope !60, !noalias !61
  call void @julia.loopinfo_marker(), !dbg !62, !julia.loopinfo !63
  %.not.not = icmp eq i64 %value_phi, 1000, !dbg !65
  %7 = add i64 %value_phi, 1, !dbg !68
  br i1 %.not.not, label %L20, label %L2, !dbg !71

LSL: has julia.loopinfo metadata with 1 operands
LSL: simd: 0 ivdep: 0
```

The Loop is not in in idiom form (e.g loop header and latch are collapsed into one BB) and `simplifycfg` eliminated the inner loop.

This issue could also occur with `@simd` and thus cause miscompilations.

### PR DESCRIPTION
**Pull request by vchuravy**
Fixes #50660

We used a loop-marker intrinsic because the LoopID used to
be dropped by optimization passes (this seems no longer true).

#50660 is an example of a miscompilation where a loop of length 1
got optimized by simplifycfg to the point where the loop-marker
is now attached to the wrong back-edge.

This PR drops the loop-marker and uses the LoopID metadata node directly.

### CODE REVIEW

File: src/llvm-simdloop.cpp [hunk 1]
```diff
@@ -283,61 +237,43 @@ static bool markLoopInfo(Module &M, Function *marker, function_ref<LoopInfo &(Fu
...
+                          LoopStandardAnalysisResults &AR, LPMUpdater &U)
 
-    if (markLoopInfo(M, loopinfo_marker, GetLI)) {
-        auto preserved = PreservedAnalyses::allInSet<CFGAnalyses>();
-        preserved.preserve<LoopAnalysis>();
+{
+    OptimizationRemarkEmitter ORE(L.getHeader()->getParent());
+    if (processLoop(L, ORE)) {
+#ifdef JL_DEBUG_BUILD
+        if (AR.MSSA)
```
[the remark this answers is not in the document]
Comment (pchintalapudi) [reply src/llvm-simdloop.cpp:247]: Loop passes are also required to preserve ScalarEvolution, which is almost certainly affected by this pass.

**Review by vtjnash — APPROVED** [review]

**Comment by vchuravy** [comment]
Currently I am miss-attaching the loop id.

File: src/codegen.cpp [hunk 2]
```diff
@@ -992,14 +992,7 @@ static const auto jl_typeof_func = new JuliaFunction<>{
             Attributes(C, {Attribute::NonNull}),
             None); },
 };
-static const auto jl_loopinfo_marker_func = new JuliaFunction<>{
```
[the remark this answers is not in the document]
Comment (vchuravy) [reply src/codegen.cpp:995]: Hm I couldn't find it xD

[the remark this answers is not in the document]
Comment (pchintalapudi) [reply src/codegen.cpp:995]: Looks like it's in `doc/src/devdocs/llvm-passes.md`, in the description of what LowerSIMDLoop does.

**Comment by vchuravy** [comment]
@pchintalapudi 

> Loop passes are also required to preserve ScalarEvolution, which is almost certainly affected by this pass.

I don't think there is a SCEV updater, but there is `forgetValue` 

So you think that in:

```
    for (chainVector::const_iterator K=chain.begin(); K!=chain.end(); ++K) {
        LLVM_DEBUG(dbgs() << "LSL: marking " << **K << "\n");
        REMARK([&]() {
            return OptimizationRemark(DEBUG_TYPE, "MarkedUnsafeAlgebra", *K)
                   << "marked unsafe algebra on " << ore::NV("Instruction", *K);
        });
        (*K)->setHasAllowReassoc(true);
        (*K)->setHasAllowContract(true);
        ++length;
    }
```

We should `forgetValue(*K)`?

**Comment by pchintalapudi** [comment]
Yes, and if we change the loop we should also call `SE->forgetLoopDispositions(L)`

**Comment by vchuravy** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by vchuravy** [comment]
@nanosoldier `runtests(vs = ":master",  configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],), vs_configuration = (buildflags = ["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],))`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/50663#issuecomment-1672469010) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/c5ee39f_vs_ebae716/report.md).

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/50663#issuecomment-1672469204) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/ec8e71e_vs_ebae716/report.html) is available.

**Comment by vchuravy** [comment]
Sigh...

```
julia: /workspace/srcdir/llvm-project/llvm/include/llvm/Support/Casting.h:566: decltype(auto) llvm::cast(const From&) [with To = llvm::MDNode; From = llvm::MDOperand]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.

[9] signal (6.-6): Aborted
in expression starting at /opt/julia/share/julia/stdlib/v1.11/Dates/test/arithmetic.jl:518
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x7f0a72f1f40e)
__assert_fail at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
_ZN4llvm23addStringMetadataToLoopEPNS_4LoopEPKcj at /opt/julia/bin/../lib/julia/libLLVM-15jl.so (unknown line)
_ZN4llvm8peelLoopEPNS_4LoopEjPNS_8LoopInfoEPNS_15ScalarEvolutionERNS_13DominatorTreeEPNS_15AssumptionCacheEb at /opt/julia/bin/../lib/julia/libLLVM-15jl.so (unknown line)
_ZL15tryToUnrollLoopPN4llvm4LoopERNS_13DominatorTreeEPNS_8LoopInfoERNS_15ScalarEvolutionERKNS_19TargetTransformInfoERNS_15AssumptionCacheERNS_25OptimizationRemarkEmitterEPNS_18BlockFrequencyInfoEPNS_18ProfileSummaryInfoEbibbNS_8OptionalIjEESK_NSJ_IbEESL_SL_SL_SL_SK_ at /opt/julia/bin/../lib/julia/libLLVM-15jl.so (unknown line)
_ZN4llvm18LoopFullUnrollPass3runERNS_4LoopERNS_15AnalysisManagerIS1_JRNS_27LoopStandardAnalysisResultsEEEES5_RNS_10LPMUpdaterE at /opt/julia/bin/../lib/julia/libLLVM-15jl.so (unknown line)
run at /source/usr/include/llvm/IR/PassManagerInternal.h:88
_ZN4llvm11PassManagerINS_4LoopENS_15AnalysisManagerIS1_JRNS_27LoopStandardAnalysisResultsEEEEJS4_RNS_10LPMUpdaterEEE13runSinglePassIS1_St10unique_ptrINS_6detail11PassConceptIS1_S5_JS4_S7_EEESt14default_deleteISD_EEEENS_8OptionalINS_17PreservedAnalysesEEERT_RT0_RS5_S4_S7_RNS_19PassInstrumentationE at /opt/julia/bin/../lib/julia/libLLVM-15jl.so (unknown line)
_ZN4llvm11PassManagerINS_4LoopENS_15AnalysisManagerIS1_JRNS_27LoopStandardAnalysisResultsEEEEJS4_RNS_10LPMUpdaterEEE24runWithoutLoopNestPassesERS1_RS5_S4_S7_ at /opt/julia/bin/../lib/julia/libLLVM-15jl.so (unknown line)
_ZN4llvm11PassManagerINS_4LoopENS_15AnalysisManagerIS1_JRNS_27LoopStandardAnalysisResultsEEEEJS4_RNS_10LPMUpdaterEEE3runERS1_RS5_S4_S7_ at /opt/julia/bin/../lib/julia/libLLVM-15jl.so (unknown line)
run at /source/usr/include/llvm/IR/PassManagerInternal.h:88
_ZN4llvm25FunctionToLoopPassAdaptor3runERNS_8FunctionERNS_15AnalysisManagerIS1_JEEE at /opt/julia/bin/../lib/julia/libLLVM-15jl.so (unknown line)
run at /source/usr/include/llvm/IR/PassManagerInternal.h:88
_ZN4llvm11PassManagerINS_8FunctionENS_15AnalysisManagerIS1_JEEEJEE3runERS1_RS3_ at /opt/julia/bin/../lib/julia/libLLVM-15jl.so (unknown line)
run at /source/usr/include/llvm/IR/PassManagerInternal.h:88
_ZN4llvm27ModuleToFunctionPassAdaptor3runERNS_6ModuleERNS_15AnalysisManagerIS1_JEEE at /opt/julia/bin/../lib/julia/libLLVM-15jl.so (unknown line)
run at /source/usr/include/llvm/IR/PassManagerInternal.h:88
_ZN4llvm11PassManagerINS_6ModuleENS_15AnalysisManagerIS1_JEEEJEE3runERS1_RS3_ at /opt/julia/bin/../lib/julia/libLLVM-15jl.so (unknown line)
```

**Comment by vchuravy** [comment]
@nanosoldier `runtests(configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],), vs_configuration = (buildflags = ["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],))`

**Comment by vchuravy** [comment]
@nanosoldier `runtests(vs = ":master",  configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],), vs_configuration = (buildflags = ["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],))`

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/50663#issuecomment-1673306994) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/853295a_vs_958da95/report.html) is available.

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/50663#issuecomment-1674142669) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/da3cc48_vs_5db443a/report.html) is available.

**Comment by IanButterworth** [comment]
FYI there are a lot of conflicts when trying to backport this to 1.9

**Pull request merged**